From b5e58bf901f0619a41dc9309573b393edf939b11 Mon Sep 17 00:00:00 2001
From: Abhilasha Manna <amanna@qti.qualcomm.com>
Date: Mon, 22 Jan 2024 18:34:50 +0530
Subject: [PATCH] fstab-generator: Honor 'verity=enabled' cmdline.

When 'verity=enabled' & 'avb-verity=enabled'is present in kernel cmdline assume verity
is enabled even if root, roothash values are not present and use
"root" DM device as root.

Signed-off-by: Abhilasha Manna<amanna@qti.qualcomm.com>
---
 src/fstab-generator/fstab-generator.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/src/fstab-generator/fstab-generator.c b/src/fstab-generator/fstab-generator.c
index 9b32383a76..c2166c38b0 100644
--- a/src/fstab-generator/fstab-generator.c
+++ b/src/fstab-generator/fstab-generator.c
@@ -52,6 +52,7 @@ static char *arg_usr_what = NULL;
 static char *arg_usr_fstype = NULL;
 static char *arg_usr_options = NULL;
 static char *arg_usr_hash = NULL;
+static char *arg_verity = NULL;
 static VolatileMode arg_volatile_mode = _VOLATILE_MODE_INVALID;
 
 STATIC_DESTRUCTOR_REGISTER(arg_root_what, freep);
@@ -62,6 +63,7 @@ STATIC_DESTRUCTOR_REGISTER(arg_usr_what, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_usr_fstype, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_usr_options, freep);
 STATIC_DESTRUCTOR_REGISTER(arg_usr_hash, freep);
+STATIC_DESTRUCTOR_REGISTER(arg_verity, freep);
 
 static int write_options(FILE *f, const char *options) {
         _cleanup_free_ char *o = NULL;
@@ -1024,6 +1026,21 @@ static int parse_proc_cmdline_item(const char *key, const char *value, void *dat
                         return 0;
 
                 return free_and_strdup_warn(&arg_root_hash, value);
+        } else if (streq(key, "verity")) {
+
+                if (proc_cmdline_value_missing(key, value))
+                        return 0;
+
+                if (free_and_strdup(&arg_verity, value) < 0)
+                        return log_oom();
+
+        } else if (streq(key, "avb-verity")) {
+
+                if (proc_cmdline_value_missing(key, value))
+                        return 0;
+
+                if (free_and_strdup(&arg_verity, value) < 0)
+                        return log_oom();
 
         } else if (streq(key, "mount.usr")) {
 
@@ -1104,6 +1121,16 @@ static int determine_device(char **what, const char *hash, const char *name) {
 }
 
 static int determine_root(void) {
+        /* If we have verity=enabled then Verity is used, and we use the "root" DM device as root. */
+        if (!isempty(arg_verity) && streq(arg_verity, "enabled")) {
+                log_info("Found a verity=enabled entry on the kernel command line.");
+                if (free_and_strdup(&arg_root_what, "/dev/mapper/root") < 0)
+                        return log_oom();
+
+                log_info("Using verity root device %s.", arg_root_what);
+                return 1;
+        }
+
         return determine_device(&arg_root_what, arg_root_hash, "root");
 }
 
-- 
2.34.1

